fix/integration-tests (#256)
canonical/wazuh-server-operator: docs/how-to/test-the-charm.md

LB_ADDR_RANGE="${LB_FIRST_ADDR}-${LB_LAST_ADDR}"

k8s set \
  load-balancer.cidrs=$LB_ADDR_RANGE \
  load-balancer.enabled=true \
  load-balancer.l2-mode=true
```

### Install kubectl

```bash
snap install kubectl --classic
```

### Install Juju

```bash
snap install juju
```

### Bootstrap the k8s controller

```bash
juju bootstrap k8s
```

### Run the pre_run script

```bash
bash -xe ~/wazuh-server-operator/tests/integration/pre_run_script.sh
```

### (Optional) Install rock dependencies

If you anticipate changing the rock container image, follow these additional
steps:

#### Install rockcraft

```bash
snap install rockcraft --classic
```

#### Install and configure Docker

```bash
apt install -y docker.io
echo '{ "insecure-registries": ["localhost:5000"] }' > /etc/docker/daemon.json
systemctl restart docker
```

#### Install skopeo

```bash
apt install -y skopeo
```

## Linting, static testing, and unit testing

This project uses `tox` for managing test environments. There are some
pre-configured environments that can be used for linting and formatting code
when you're preparing contributions to the charm:

> ```bash
> tox run -e fmt        # update your code according to linting rules
> tox run -e lint       # code style
> tox run -e static     # other checks such as `bandit` for security issues.
> tox run -e unit       # unit tests
> tox                   # runs 'format', 'lint', 'static' and 'unit' environments
> ```

Reviewer: To avoid the Vale warning, I recommend using the following formatting (see [CONTRIBUTING.md in platform-engineering-charm-template](https://github.com/canonical/platform-engineering-charm-template/blob/main/CONTRIBUTING.md?plain=1)):
````suggestion
* ``tox``: Executes all of the basic checks and tests (``lint``, ``unit``, ``static``, and ``format``).
* ``tox run -e fmt``: Update your code according to linting rules.
* ``tox run -e lint``: Runs a range of static code analysis to check the code.
* ``tox run -e static``: Runs other checks such as ``bandit`` for security issues.
* ``tox run -e unit``: Runs unit tests.
````

## Integration testing

Integration testing is a multi-step process that requires:

1. Building the charm
1. (Optionally) building the rock
1. Running the tests

#### Build the charm

```bash
cd ~/wazuh-server-operator
charmcraft pack
```

#### Optional - build the rock

If you have not made any changes to the rock, you do not need to rebuild it.

> The Github integration test workflow builds and uploads the rock to `ghcr.io`

Reviewer: ```suggestion
The GitHub integration test workflow builds and uploads the rock to `ghcr.io`
```

> for its own tests. If you haven't changed the rock since the last Github action

Reviewer: ```suggestion
for its own tests. If you haven't changed the rock since the last GitHub action
```

rockcraft pack
```

Upload the rock into a local registry:

```bash
docker run -d -p 5000:5000 --restart always --name registry registry:2
skopeo --insecure-policy copy --dest-tls-verify=false \
    oci-archive:wazuh-server_1.0_amd64.rock \
    docker://localhost:5000/wazuh-server:latest
```

#### Set the container location

If you rebuilt the rock:

```bash
export IMAGE_URL="localhost:5000/wazuh-server:latest"
```

If you did not rebuild the rock:

```bash
# find latest tag here:
# https://github.com/canonical/wazuh-server-operator/pkgs/container/wazuh-server
export IMAGE_URL="ghcr.io/canonical/wazuh-server:a063aca515693126206e4dfa6ba6eba4bac43698-_1.0_amd64"
```

#### Run tests

##### With three `wazuh-indexer` nodes

_minimum 32 GB of RAM suggested_

```bash
tox run -e integration -- \
    --charm-file=wazuh-server_ubuntu-22.04-amd64.charm \
    --wazuh-server-image $IMAGE_URL \
    --controller k8s --model test-wazuh
```

##### With a single `wazuh-indexer` node

```bash
tox run -e integration -- \
    --charm-file=wazuh-server_ubuntu-22.04-amd64.charm \
    --wazuh-server-image $IMAGE_URL \
    --single-node-indexer \
    --controller k8s --model test-wazuh
```

> ##### To reuse environments

Reviewer: ```suggestion
##### Reuse environments
```
Nitpick

```bash
k8s enable load-balancer

IPADDR=$(ip -4 -j route get 2.2.2.2 | jq -r '.[] | .prefsrc')
LB_FIRST_ADDR="$(echo "${IPADDR}" | awk -F'.' '{print $1,$2,$3,225}' OFS='.')"
LB_LAST_ADDR="$(echo "${IPADDR}" | awk -F'.' '{print $1,$2,$3,250}' OFS='.')"
LB_ADDR_RANGE="${LB_FIRST_ADDR}-${LB_LAST_ADDR}"

k8s set \
  load-balancer.cidrs=$LB_ADDR_RANGE \
  load-balancer.enabled=true \
  load-balancer.l2-mode=true
```

### Install kubectl

```bash
snap install kubectl --classic
```

### Install Juju

```bash
snap install juju
```

### Bootstrap the Kubernetes controller

```bash
juju bootstrap k8s
```

### Run the pre-run script

```bash
bash -xe ~/wazuh-server-operator/tests/integration/pre_run_script.sh
```

### (Optional) Install rock dependencies

If you anticipate changing the rock container image, follow these additional
steps:

#### Install Rockcraft

```bash
snap install rockcraft --classic
```

#### Install and configure Docker

```bash
apt install -y docker.io
echo '{ "insecure-registries": ["localhost:5000"] }' > /etc/docker/daemon.json
systemctl restart docker
```

#### Install skopeo

```bash
apt install -y skopeo
```

> ## Linting, static testing, and unit testing

Reviewer: ```suggestion
## Testing
```
Nitpick, just to be more concise

This project uses `tox` for managing test environments. There are some
pre-configured environments that can be used for linting and formatting code
when you're preparing contributions to the charm:

* ``tox``: Executes all of the basic checks and tests (``lint``, ``unit``, ``static``, and ``format``).
* ``tox run -e fmt``: Update your code according to linting rules.
* ``tox run -e lint``: Runs a range of static code analysis to check the code.
* ``tox run -e static``: Runs other checks such as ``bandit`` for security issues.
* ``tox run -e unit``: Runs unit tests.

## Integration testing

Integration testing is a multi-step process that requires:

1. Building the charm
1. (Optionally) building the rock
1. Running the tests

#### Build the charm

```bash
cd ~/wazuh-server-operator
charmcraft pack
```

> #### Optional - build the rock

Reviewer: ```suggestion
#### (Optional) Build the rock
```
Nitpick, just to be consistent with the previous `(Optional)` header style
